File: sample/go/18_mini_language_interpreter.go
package main

type TokenLike interface {
}

type ExprNodeLike interface {
}

type StmtNodeLike interface {
}

type ParserLike interface {
    new_expr_nodes() []any
    current_token() TokenLike
    previous_token() TokenLike
    peek_kind() string
    match(kind string) bool
    expect(kind string) TokenLike
    skip_newlines()
    add_expr(node ExprNodeLike) int64
    parse_program() []any
    parse_stmt() StmtNodeLike
    parse_expr() int64
    parse_add() int64
    parse_mul() int64
    parse_unary() int64
    parse_primary() int64
}


func __pytra_is_Token(v any) bool {
    _, ok := v.(*Token)
    return ok
}

func __pytra_as_Token(v any) *Token {
    if t, ok := v.(*Token); ok {
        return t
    }
    return nil
}

func __pytra_is_ExprNode(v any) bool {
    _, ok := v.(*ExprNode)
    return ok
}

func __pytra_as_ExprNode(v any) *ExprNode {
    if t, ok := v.(*ExprNode); ok {
        return t
    }
    return nil
}

func __pytra_is_StmtNode(v any) bool {
    _, ok := v.(*StmtNode)
    return ok
}

func __pytra_as_StmtNode(v any) *StmtNode {
    if t, ok := v.(*StmtNode); ok {
        return t
    }
    return nil
}

func __pytra_is_Parser(v any) bool {
    _, ok := v.(*Parser)
    return ok
}

func __pytra_as_Parser(v any) *Parser {
    if t, ok := v.(*Parser); ok {
        return t
    }
    return nil
}

type Token struct {
    kind string
    text string
    pos int64
    number_value int64
}

func NewToken(kind string, text string, pos int64, number_value int64) *Token {
    self := &Token{}
    self.kind = kind
    self.text = text
    self.pos = pos
    self.number_value = number_value
    return self
}

type ExprNode struct {
    kind string
    value int64
    name string
    op string
    left int64
    right int64
    kind_tag int64
    op_tag int64
}

func NewExprNode(kind string, value int64, name string, op string, left int64, right int64, kind_tag int64, op_tag int64) *ExprNode {
    self := &ExprNode{}
    self.kind = kind
    self.value = value
    self.name = name
    self.op = op
    self.left = left
    self.right = right
    self.kind_tag = kind_tag
    self.op_tag = op_tag
    return self
}

type StmtNode struct {
    kind string
    name string
    expr_index int64
    kind_tag int64
}

func NewStmtNode(kind string, name string, expr_index int64, kind_tag int64) *StmtNode {
    self := &StmtNode{}
    self.kind = kind
    self.name = name
    self.expr_index = expr_index
    self.kind_tag = kind_tag
    return self
}

type Parser struct {
    tokens []any
    pos int64
    expr_nodes []any
}

func NewParser(tokens []any) *Parser {
    self := &Parser{}
    self.Init(tokens)
    return self
}

func (self *Parser) new_expr_nodes() []any {
    return __pytra_as_list([]any{})
}

func (self *Parser) Init(tokens []any) {
    self.tokens = tokens
    self.pos = int64(0)
    self.expr_nodes = self.new_expr_nodes()
}

func (self *Parser) current_token() TokenLike {
    return __pytra_get_index(self.tokens, self.pos)
}

func (self *Parser) previous_token() TokenLike {
    return __pytra_get_index(self.tokens, (self.pos - int64(1)))
}

func (self *Parser) peek_kind() string {
    return __pytra_str(self.current_token().kind)
}

func (self *Parser) match(kind string) bool {
    if (__pytra_str(self.peek_kind()) == __pytra_str(kind)) {
        self.pos += int64(1)
        return __pytra_truthy(true)
    }
    return __pytra_truthy(false)
}

func (self *Parser) expect(kind string) TokenLike {
    var token TokenLike = self.current_token()
    if (__pytra_str(token.kind) != __pytra_str(kind)) {
        panic(__pytra_str(((__pytra_str((__pytra_str((__pytra_str((__pytra_str("parse error at pos=") + __pytra_str(__pytra_str(token.pos)))) + __pytra_str(", expected="))) + __pytra_str(kind))) + __pytra_str(", got=")) + token.kind)))
    }
    self.pos += int64(1)
    return token
}

func (self *Parser) skip_newlines() {
    for self.match("NEWLINE") {
        _ = 0
    }
}

func (self *Parser) add_expr(node ExprNodeLike) int64 {
    self.expr_nodes = append(self.expr_nodes, node)
    return (__pytra_len(self.expr_nodes) - int64(1))
}

func (self *Parser) parse_program() []any {
    var stmts []any = __pytra_as_list([]any{})
    self.skip_newlines()
    for (__pytra_str(self.peek_kind()) != __pytra_str("EOF")) {
        var stmt StmtNodeLike = self.parse_stmt()
        stmts = append(stmts, stmt)
        self.skip_newlines()
    }
    return __pytra_as_list(stmts)
}

func (self *Parser) parse_stmt() StmtNodeLike {
    if self.match("LET") {
        var let_name string = __pytra_str(self.expect("IDENT").text)
        self.expect("EQUAL")
        var let_expr_index int64 = self.parse_expr()
        return NewStmtNode("let", let_name, let_expr_index, int64(1))
    }
    if self.match("PRINT") {
        var print_expr_index int64 = self.parse_expr()
        return NewStmtNode("print", "", print_expr_index, int64(3))
    }
    var assign_name string = __pytra_str(self.expect("IDENT").text)
    self.expect("EQUAL")
    var assign_expr_index int64 = self.parse_expr()
    return NewStmtNode("assign", assign_name, assign_expr_index, int64(2))
}

func (self *Parser) parse_expr() int64 {
    return self.parse_add()
}

func (self *Parser) parse_add() int64 {
    var left int64 = self.parse_mul()
    for true {
        if self.match("PLUS") {
            var right int64 = self.parse_mul()
            left = self.add_expr(NewExprNode("bin", int64(0), "", "+", left, right, int64(3), int64(1)))
            continue
        }
        if self.match("MINUS") {
            var right int64 = self.parse_mul()
            left = self.add_expr(NewExprNode("bin", int64(0), "", "-", left, right, int64(3), int64(2)))
            continue
        }
        break
    }
    return left
}

func (self *Parser) parse_mul() int64 {
    var left int64 = self.parse_unary()
    for true {
        if self.match("STAR") {
            var right int64 = self.parse_unary()
            left = self.add_expr(NewExprNode("bin", int64(0), "", "*", left, right, int64(3), int64(3)))
            continue
        }
        if self.match("SLASH") {
            var right int64 = self.parse_unary()
            left = self.add_expr(NewExprNode("bin", int64(0), "", "/", left, right, int64(3), int64(4)))
            continue
        }
        break
    }
    return left
}

func (self *Parser) parse_unary() int64 {
    if self.match("MINUS") {
        var child int64 = self.parse_unary()
        return self.add_expr(NewExprNode("neg", int64(0), "", "", child, (-int64(1)), int64(4), int64(0)))
    }
    return self.parse_primary()
}

func (self *Parser) parse_primary() int64 {
    if self.match("NUMBER") {
        var token_num TokenLike = self.previous_token()
        return self.add_expr(NewExprNode("lit", token_num.number_value, "", "", (-int64(1)), (-int64(1)), int64(1), int64(0)))
    }
    if self.match("IDENT") {
        var token_ident TokenLike = self.previous_token()
        return self.add_expr(NewExprNode("var", int64(0), token_ident.text, "", (-int64(1)), (-int64(1)), int64(2), int64(0)))
    }
    if self.match("LPAREN") {
        var expr_index int64 = self.parse_expr()
        self.expect("RPAREN")
        return expr_index
    }
    var t TokenLike = self.current_token()
    panic(__pytra_str(((__pytra_str((__pytra_str("primary parse error at pos=") + __pytra_str(__pytra_str(t.pos)))) + __pytra_str(" got=")) + t.kind)))
    return 0
}

func tokenize(lines []any) []any {
    var single_char_token_tags map[any]any = __pytra_as_dict(map[any]any{})
    var single_char_token_kinds []any = __pytra_as_list([]any{"PLUS", "MINUS", "STAR", "SLASH", "LPAREN", "RPAREN", "EQUAL"})
    var tokens []any = __pytra_as_list([]any{})
    __iter_0 := __pytra_as_list(__pytra_enumerate(lines))
    for __i_1 := int64(0); __i_1 < int64(len(__iter_0)); __i_1 += 1 {
        __it_2 := __iter_0[__i_1]
        __tuple_3 := __pytra_as_list(__it_2)
        var line_index int64 = __pytra_int(__tuple_3[0])
        _ = line_index
        var source string = __pytra_str(__tuple_3[1])
        _ = source
        var i int64 = int64(0)
        var n int64 = __pytra_len(source)
        for (i < n) {
            var ch string = __pytra_str(__pytra_str(__pytra_get_index(source, i)))
            if (__pytra_str(ch) == __pytra_str(" ")) {
                i += int64(1)
                continue
            }
            var single_tag int64 = single_char_token_tags.get(ch, int64(0))
            if (single_tag > int64(0)) {
                tokens = append(tokens, NewToken(__pytra_str(__pytra_get_index(single_char_token_kinds, (single_tag - int64(1)))), ch, i, int64(0)))
                i += int64(1)
                continue
            }
            if __pytra_truthy(__pytra_isdigit(ch)) {
                var start int64 = i
                for ((i < n) && __pytra_truthy(__pytra_isdigit(__pytra_str(__pytra_get_index(source, i))))) {
                    i += int64(1)
                }
                var text string = __pytra_str(__pytra_slice(source, start, i))
                tokens = append(tokens, NewToken("NUMBER", text, start, __pytra_int(text)))
                continue
            }
            if (__pytra_truthy(__pytra_isalpha(ch)) || (__pytra_str(ch) == __pytra_str("_"))) {
                var start int64 = i
                for ((i < n) && ((__pytra_truthy(__pytra_isalpha(__pytra_str(__pytra_get_index(source, i)))) || (__pytra_str(__pytra_str(__pytra_get_index(source, i))) == __pytra_str("_"))) || __pytra_truthy(__pytra_isdigit(__pytra_str(__pytra_get_index(source, i)))))) {
                    i += int64(1)
                }
                var text string = __pytra_str(__pytra_slice(source, start, i))
                if (__pytra_str(text) == __pytra_str("let")) {
                    tokens = append(tokens, NewToken("LET", text, start, int64(0)))
                } else {
                    if (__pytra_str(text) == __pytra_str("print")) {
                        tokens = append(tokens, NewToken("PRINT", text, start, int64(0)))
                    } else {
                        tokens = append(tokens, NewToken("IDENT", text, start, int64(0)))
                    }
                }
                continue
            }
            panic(__pytra_str((__pytra_str((__pytra_str((__pytra_str((__pytra_str((__pytra_str("tokenize error at line=") + __pytra_str(__pytra_str(line_index)))) + __pytra_str(" pos="))) + __pytra_str(__pytra_str(i)))) + __pytra_str(" ch="))) + __pytra_str(ch))))
        }
        tokens = append(tokens, NewToken("NEWLINE", "", n, int64(0)))
    }
    tokens = append(tokens, NewToken("EOF", "", __pytra_len(lines), int64(0)))
    return __pytra_as_list(tokens)
}

func eval_expr(expr_index int64, expr_nodes []any, env map[any]any) int64 {
    var node ExprNodeLike = __pytra_get_index(expr_nodes, expr_index)
    if (__pytra_int(node.kind_tag) == int64(1)) {
        return __pytra_int(node.value)
    }
    if (__pytra_int(node.kind_tag) == int64(2)) {
        if (!(__pytra_contains(env, node.name))) {
            panic(__pytra_str(("undefined variable: " + node.name)))
        }
        return __pytra_int(__pytra_get_index(env, node.name))
    }
    if (__pytra_int(node.kind_tag) == int64(4)) {
        return (-eval_expr(node.left, expr_nodes, env))
    }
    if (__pytra_int(node.kind_tag) == int64(3)) {
        var lhs int64 = eval_expr(node.left, expr_nodes, env)
        var rhs int64 = eval_expr(node.right, expr_nodes, env)
        if (__pytra_int(node.op_tag) == int64(1)) {
            return (lhs + rhs)
        }
        if (__pytra_int(node.op_tag) == int64(2)) {
            return (lhs - rhs)
        }
        if (__pytra_int(node.op_tag) == int64(3)) {
            return (lhs * rhs)
        }
        if (__pytra_int(node.op_tag) == int64(4)) {
            if (rhs == int64(0)) {
                panic(__pytra_str("division by zero"))
            }
            return __pytra_int((lhs / rhs))
        }
        panic(__pytra_str(("unknown operator: " + node.op)))
    }
    panic(__pytra_str(("unknown node kind: " + node.kind)))
    return 0
}

func execute(stmts []any, expr_nodes []any, trace bool) int64 {
    var env map[any]any = __pytra_as_dict(map[any]any{})
    var checksum int64 = int64(0)
    var printed int64 = int64(0)
    __iter_0 := __pytra_as_list(stmts)
    for __i_1 := int64(0); __i_1 < int64(len(__iter_0)); __i_1 += 1 {
        var stmt StmtNodeLike = __iter_0[__i_1]
        if (__pytra_int(stmt.kind_tag) == int64(1)) {
            __pytra_set_index(env, stmt.name, eval_expr(stmt.expr_index, expr_nodes, env))
            continue
        }
        if (__pytra_int(stmt.kind_tag) == int64(2)) {
            if (!(__pytra_contains(env, stmt.name))) {
                panic(__pytra_str(("assign to undefined variable: " + stmt.name)))
            }
            __pytra_set_index(env, stmt.name, eval_expr(stmt.expr_index, expr_nodes, env))
            continue
        }
        var value int64 = eval_expr(stmt.expr_index, expr_nodes, env)
        if trace {
            __pytra_print(value)
        }
        var norm int64 = (value % int64(1000000007))
        if (norm < int64(0)) {
            norm += int64(1000000007)
        }
        checksum = (((checksum * int64(131)) + norm) % int64(1000000007))
        printed += int64(1)
    }
    if trace {
        __pytra_print("printed:", printed)
    }
    return checksum
}

func build_benchmark_source(var_count int64, loops int64) []any {
    var lines []any = __pytra_as_list([]any{})
    for i := int64(0); i < var_count; i += 1 {
        lines = append(lines, (__pytra_str((__pytra_str((__pytra_str("let v") + __pytra_str(__pytra_str(i)))) + __pytra_str(" = "))) + __pytra_str(__pytra_str((i + int64(1))))))
    }
    for i := int64(0); i < loops; i += 1 {
        var x int64 = (i % var_count)
        var y int64 = ((i + int64(3)) % var_count)
        var c1 int64 = ((i % int64(7)) + int64(1))
        var c2 int64 = ((i % int64(11)) + int64(2))
        lines = append(lines, (__pytra_str((__pytra_str((__pytra_str((__pytra_str((__pytra_str((__pytra_str((__pytra_str((__pytra_str((__pytra_str("v") + __pytra_str(__pytra_str(x)))) + __pytra_str(" = (v"))) + __pytra_str(__pytra_str(x)))) + __pytra_str(" * "))) + __pytra_str(__pytra_str(c1)))) + __pytra_str(" + v"))) + __pytra_str(__pytra_str(y)))) + __pytra_str(" + 10000) / "))) + __pytra_str(__pytra_str(c2))))
        if ((i % int64(97)) == int64(0)) {
            lines = append(lines, (__pytra_str("print v") + __pytra_str(__pytra_str(x))))
        }
    }
    lines = append(lines, "print (v0 + v1 + v2 + v3)")
    return __pytra_as_list(lines)
}

func run_demo() {
    var demo_lines []any = __pytra_as_list([]any{})
    demo_lines = append(demo_lines, "let a = 10")
    demo_lines = append(demo_lines, "let b = 3")
    demo_lines = append(demo_lines, "a = (a + b) * 2")
    demo_lines = append(demo_lines, "print a")
    demo_lines = append(demo_lines, "print a / b")
    var tokens []any = __pytra_as_list(tokenize(demo_lines))
    var parser ParserLike = NewParser(tokens)
    var stmts []any = __pytra_as_list(parser.parse_program())
    var checksum int64 = execute(stmts, parser.expr_nodes, true)
    __pytra_print("demo_checksum:", checksum)
}

func run_benchmark() {
    var source_lines []any = __pytra_as_list(build_benchmark_source(int64(32), int64(120000)))
    var start float64 = __pytra_perf_counter()
    var tokens []any = __pytra_as_list(tokenize(source_lines))
    var parser ParserLike = NewParser(tokens)
    var stmts []any = __pytra_as_list(parser.parse_program())
    var checksum int64 = execute(stmts, parser.expr_nodes, false)
    var elapsed float64 = (__pytra_perf_counter() - start)
    __pytra_print("token_count:", __pytra_len(tokens))
    __pytra_print("expr_count:", __pytra_len(parser.expr_nodes))
    __pytra_print("stmt_count:", __pytra_len(stmts))
    __pytra_print("checksum:", checksum)
    __pytra_print("elapsed_sec:", elapsed)
}

func __pytra_main() {
    run_demo()
    run_benchmark()
}

func main() {
    __pytra_main()
}

File: sample/nim/18_mini_language_interpreter.nim
include "py_runtime.nim"

import std/os, std/times, std/tables, std/strutils, std/math

type Token* = ref object
  kind*: string
  text*: string
  pos*: int
  number_value*: int

type ExprNode* = ref object
  kind*: string
  value*: int
  name*: string
  op*: string
  left*: int
  right*: int
  kind_tag*: int
  op_tag*: int

type StmtNode* = ref object
  kind*: string
  name*: string
  expr_index*: int
  kind_tag*: int

proc tokenize*(lines: seq[string]): auto =
  var single_char_token_tags: Table[string, int] = { "+": 1, "-": 2, "*": 3, "/": 4, "(": 5, ")": 6, "=": 7 }.toTable
  var single_char_token_kinds: seq[string] = @["PLUS", "MINUS", "STAR", "SLASH", "LPAREN", "RPAREN", "EQUAL"]
  var tokens: seq[Token] = @[]
  for it in enumerate(lines):
    var i: int = 0
    var n: int = source.len
    while (i < n):
      var ch: string = source[i]
      if (ch == " "):
        i += 1
        discard `continue`
      var single_tag: int = /* unknown expr Unbox */
      if (single_tag > 0):
        tokens.add(newToken(single_char_token_kinds[(single_tag - 1)], ch, i, 0))
        i += 1
        discard `continue`
      if py_truthy(ch.isdigit()):
        var start: int = i
        while py_truthy(((i < n) {op} py_truthy(source[i].isdigit()))):
          i += 1
        var text: string = source[start ..< i]
        tokens.add(newToken("NUMBER", text, start, int(text)))
        discard `continue`
      if py_truthy((py_truthy(ch.isalpha()) {op} (ch == "_"))):
        start = i
        while py_truthy(((i < n) {op} py_truthy((py_truthy((py_truthy(source[i].isalpha()) {op} (source[i] == "_"))) {op} py_truthy(source[i].isdigit()))))):
          i += 1
        text = source[start ..< i]
        if (text == "let"):
          tokens.add(newToken("LET", text, start, 0))
        elif (text == "print"):
          tokens.add(newToken("PRINT", text, start, 0))
        else:
          tokens.add(newToken("IDENT", text, start, 0))
        discard `continue`
      raise newException(Exception, RuntimeError(((((("tokenize error at line=" & $( line_index )) & " pos=") & $( i )) & " ch=") & ch)))
    tokens.add(newToken("NEWLINE", "", n, 0))
  tokens.add(newToken("EOF", "", lines.len, 0))
  return tokens

type Parser* = ref object
  discard

proc new_expr_nodes*(self: Parser): auto =
  return @[]

proc newParser*(tokens: seq[Token]): Parser =
  new(result)
  result.tokens = tokens # seq[Token]
  result.pos = 0 # int
  result.expr_nodes = result.new_expr_nodes() # seq[ExprNode]

proc current_token*(self: Parser): auto =
  return self.tokens[self.pos]

proc previous_token*(self: Parser): auto =
  return self.tokens[(self.pos - 1)]

proc peek_kind*(self: Parser): auto =
  return self.current_token().kind

proc match*(self: Parser, kind: string): auto =
  if (self.peek_kind() == kind):
    self.pos += 1
    return true
  return false

proc expect*(self: Parser, kind: string): auto =
  var token: Token = self.current_token()
  if (token.kind != kind):
    raise newException(Exception, RuntimeError(((((("parse error at pos=" & /* unknown expr ObjStr */) & ", expected=") & kind) & ", got=") + token.kind)))
  self.pos += 1
  return token

proc skip_newlines*(self: Parser): auto =
  while py_truthy(self.match("NEWLINE")):
    discard

proc add_expr*(self: Parser, node: ExprNode): auto =
  self.expr_nodes.add(node)
  return (self.expr_nodes.len - 1)

proc parse_program*(self: Parser): auto =
  var stmts: seq[StmtNode] = @[]
  discard self.skip_newlines()
  while (self.peek_kind() != "EOF"):
    var stmt: StmtNode = self.parse_stmt()
    stmts.add(stmt)
    discard self.skip_newlines()
  return stmts

proc parse_stmt*(self: Parser): auto =
  if py_truthy(self.match("LET")):
    var let_name: string = /* unknown expr Unbox */
    discard self.expect("EQUAL")
    var let_expr_index: int = self.parse_expr()
    return newStmtNode("let", let_name, let_expr_index, 1)
  if py_truthy(self.match("PRINT")):
    var print_expr_index: int = self.parse_expr()
    return newStmtNode("print", "", print_expr_index, 3)
  var assign_name: string = /* unknown expr Unbox */
  discard self.expect("EQUAL")
  var assign_expr_index: int = self.parse_expr()
  return newStmtNode("assign", assign_name, assign_expr_index, 2)

proc parse_expr*(self: Parser): auto =
  return self.parse_add()

proc parse_add*(self: Parser): auto =
  var left: int = self.parse_mul()
  while true:
    if py_truthy(self.match("PLUS")):
      var right: int = self.parse_mul()
      left = self.add_expr(newExprNode("bin", 0, "", "+", left, right, 3, 1))
      discard `continue`
    if py_truthy(self.match("MINUS")):
      right = self.parse_mul()
      left = self.add_expr(newExprNode("bin", 0, "", "-", left, right, 3, 2))
      discard `continue`
    discard `break`
  return left

proc parse_mul*(self: Parser): auto =
  var left: int = self.parse_unary()
  while true:
    if py_truthy(self.match("STAR")):
      var right: int = self.parse_unary()
      left = self.add_expr(newExprNode("bin", 0, "", "*", left, right, 3, 3))
      discard `continue`
    if py_truthy(self.match("SLASH")):
      right = self.parse_unary()
      left = self.add_expr(newExprNode("bin", 0, "", "/", left, right, 3, 4))
      discard `continue`
    discard `break`
  return left

proc parse_unary*(self: Parser): auto =
  if py_truthy(self.match("MINUS")):
    var child: int = self.parse_unary()
    return self.add_expr(newExprNode("neg", 0, "", "", child, (-1), 4, 0))
  return self.parse_primary()

proc parse_primary*(self: Parser): auto =
  if py_truthy(self.match("NUMBER")):
    var token_num: Token = self.previous_token()
    return self.add_expr(newExprNode("lit", token_num.number_value, "", "", (-1), (-1), 1, 0))
  if py_truthy(self.match("IDENT")):
    var token_ident: Token = self.previous_token()
    return self.add_expr(newExprNode("var", 0, token_ident.text, "", (-1), (-1), 2, 0))
  if py_truthy(self.match("LPAREN")):
    var expr_index: int = self.parse_expr()
    discard self.expect("RPAREN")
    return expr_index
  var t = self.current_token()
  raise newException(Exception, RuntimeError(((("primary parse error at pos=" & /* unknown expr ObjStr */) & " got=") + t.kind)))

proc eval_expr*(expr_index: int, expr_nodes: seq[ExprNode], env: Table[string, int]): auto =
  var node: ExprNode = expr_nodes[expr_index]
  if (node.kind_tag == 1):
    return node.value
  if (node.kind_tag == 2):
    if py_truthy((not (node.name == env))):
      raise newException(Exception, RuntimeError(("undefined variable: " + node.name)))
    return env[node.name]
  if (node.kind_tag == 4):
    return (-eval_expr(node.left, expr_nodes, env))
  if (node.kind_tag == 3):
    var lhs: int = eval_expr(node.left, expr_nodes, env)
    var rhs: int = eval_expr(node.right, expr_nodes, env)
    if (node.op_tag == 1):
      return (lhs + rhs)
    if (node.op_tag == 2):
      return (lhs - rhs)
    if (node.op_tag == 3):
      return (lhs * rhs)
    if (node.op_tag == 4):
      if (rhs == 0):
        raise newException(Exception, RuntimeError("division by zero"))
      return (lhs div rhs)
    raise newException(Exception, RuntimeError(("unknown operator: " + node.op)))
  raise newException(Exception, RuntimeError(("unknown node kind: " + node.kind)))

proc execute*(stmts: seq[StmtNode], expr_nodes: seq[ExprNode], trace: bool): auto =
  var env: Table[string, int] = {  }.toTable
  var checksum: int = 0
  var printed: int = 0
  for stmt in stmts:
    if (stmt.kind_tag == 1):
      env[stmt.name] = eval_expr(stmt.expr_index, expr_nodes, env)
      discard `continue`
    if (stmt.kind_tag == 2):
      if py_truthy((not (stmt.name == env))):
        raise newException(Exception, RuntimeError(("assign to undefined variable: " + stmt.name)))
      env[stmt.name] = eval_expr(stmt.expr_index, expr_nodes, env)
      discard `continue`
    var value: int = eval_expr(stmt.expr_index, expr_nodes, env)
    if py_truthy(trace):
      echo value
    var norm: int = py_mod(value, 1000000007)
    if (norm < 0):
      norm += 1000000007
    checksum = py_mod(((checksum * 131) + norm), 1000000007)
    printed += 1
  if py_truthy(trace):
    echo "printed:", printed
  return checksum

proc build_benchmark_source*(var_count: int, loops: int): auto =
  var lines: seq[string] = @[]
  for i in 0 ..< var_count:
    lines.add(((("let v" & $( i )) & " = ") & $( (i + 1) )))
  for i in 0 ..< loops:
    var x: int = py_mod(i, var_count)
    var y: int = py_mod((i + 3), var_count)
    var c1: int = (py_mod(i, 7) + 1)
    var c2: int = (py_mod(i, 11) + 2)
    lines.add(((((((((("v" & $( x )) & " = (v") & $( x )) & " * ") & $( c1 )) & " + v") & $( y )) & " + 10000) / ") & $( c2 )))
    if (py_mod(i, 97) == 0):
      lines.add(("print v" & $( x )))
  lines.add("print (v0 + v1 + v2 + v3)")
  return lines

proc run_demo*(): auto =
  var demo_lines: seq[string] = @[]
  demo_lines.add("let a = 10")
  demo_lines.add("let b = 3")
  demo_lines.add("a = (a + b) * 2")
  demo_lines.add("print a")
  demo_lines.add("print a / b")
  var tokens: seq[Token] = tokenize(demo_lines)
  var parser: Parser = newParser(tokens)
  var stmts: seq[StmtNode] = parser.parse_program()
  var checksum: int = execute(stmts, parser.expr_nodes, true)
  echo "demo_checksum:", checksum

proc run_benchmark*(): auto =
  var source_lines: seq[string] = build_benchmark_source(32, 120000)
  var start: float = epochTime()
  var tokens: seq[Token] = tokenize(source_lines)
  var parser: Parser = newParser(tokens)
  var stmts: seq[StmtNode] = parser.parse_program()
  var checksum: int = execute(stmts, parser.expr_nodes, false)
  var elapsed: float = (epochTime() - start)
  echo "token_count:", tokens.len
  echo "expr_count:", /* unknown expr ObjLen */
  echo "stmt_count:", stmts.len
  echo "checksum:", checksum
  echo "elapsed_sec:", elapsed

proc v_pytra_main*(): auto =
  run_demo()
  run_benchmark()


if isMainModule:
  discard main()
